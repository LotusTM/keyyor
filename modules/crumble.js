const _ = require('lodash')
const { extname } = require('path')

/**
 * Explodes string path into array breadcrumb
 * @param  {string} to Relative or absolute path
 * @return {array}     Array, which consists of path's fragments
 * @example crumble('/url/to/index.html') -> `['url', 'to']`
 * @example crumble('/url/to/404.html') -> `['url', 'to', '404']`
 * @example crumble('/') -> `['index']`
*/
module.exports = (to) => {
  if (to === '/') return ['index']

  const breadcrumb = _.chain(to)
    .trimStart('/')
    .trimEnd('/')
    .replace(new RegExp(`${extname(to)}$`, 'ig'), '')
    .split('/')
    .value()

  if ((breadcrumb.length >= 2) && (_.last(breadcrumb) === 'index')) {
    breadcrumb.pop()
  }

  return breadcrumb
}
