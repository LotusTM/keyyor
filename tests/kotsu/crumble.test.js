/* eslint-env jest */

import crumble from '../../modules/crumble'

describe('Crumble function', () => {
  it('should explode absolute path into array', () => {
    expect(crumble('/this/is/testing/path')).toMatchSnapshot()
  })

  it('should explode relative path into array', () => {
    expect(crumble('this/is/testing/path')).toMatchSnapshot()
  })

  it('should return array with single value `index` for root', () => {
    expect(crumble('/')).toMatchSnapshot()
  })

  it('should return array with single value `index` if `index` is the only part of absolute path', () => {
    expect(crumble('/index')).toMatchSnapshot()
  })

  it('should return array with single value `index` if `index` is the only part of relative path', () => {
    expect(crumble('index')).toMatchSnapshot()
  })

  it('should strip file extensions', () => {
    expect(crumble('this/is/testing/path.html')).toMatchSnapshot()
  })

  it('should strip last `index` from path', () => {
    expect(crumble('this/is/testing/path/index')).toMatchSnapshot()
  })

  it('should strip last `index` with extension from path', () => {
    expect(crumble('this/is/testing/path/index.html')).toMatchSnapshot()
  })
})
