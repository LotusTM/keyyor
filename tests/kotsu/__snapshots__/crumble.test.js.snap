// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crumble function should explode absolute path into array 1`] = `
Array [
  "this",
  "is",
  "testing",
  "path",
]
`;

exports[`Crumble function should explode relative path into array 1`] = `
Array [
  "this",
  "is",
  "testing",
  "path",
]
`;

exports[`Crumble function should return array with single value \`index\` for root 1`] = `
Array [
  "index",
]
`;

exports[`Crumble function should return array with single value \`index\` if \`index\` is the only part of absolute path 1`] = `
Array [
  "index",
]
`;

exports[`Crumble function should return array with single value \`index\` if \`index\` is the only part of relative path 1`] = `
Array [
  "index",
]
`;

exports[`Crumble function should strip file extensions 1`] = `
Array [
  "this",
  "is",
  "testing",
  "path",
]
`;

exports[`Crumble function should strip last \`index\` from path 1`] = `
Array [
  "this",
  "is",
  "testing",
  "path",
]
`;

exports[`Crumble function should strip last \`index\` with extension from path 1`] = `
Array [
  "this",
  "is",
  "testing",
  "path",
]
`;
